Fix close websocket on web and a few related cleanups

klondikedragon opened the pull request:
Fixes https://github.com/zino-hofmann/graphql-flutter/issues/968 by using `ws_status.normalClosure` instead of `ws_status.goingAway`

Some additional issues were encountered based on some manual testing of graceful and forced websocket disconnects under different conditions on the web platform, so this PR also fixes these as well:
* Solves a race condition when the websocket was programmatically closed where if the `onCancel` callback was called before the socket connection state was set to `notConnected`, it was trying to send a "shutdown" message on the now-closed socket, causing an exception. This is solved by always setting the connection state to `notConnected` before the websocket is programmatically closed.
* A race condition existed in the dispose call if onConnectionLost was called before the connectionStateController was caused, and autoReconnect was true -- it could lead to a situation where it was trying to reconnect the socket during the dispose call after it just had shut down the websocket. This is solved by the combination of the solution to the previous race condition, as well as a new variable that explicitly indicates that the `SocketClient` has been disposed and it should no longer attempt to reconnect or do other similar work.
* Fixed a bug where the connection would not automatically reconnect even if autoReconnect was true if it was the server that had forcefully disconnected the websocket (e.g., the GraphQL server had shut down). This was fixed by setting `cancelOnError` on the `_messages` stream to false, so that it will properly call the `onDone` callback (triggers `onConnectionLost`, which triggers the reconnect). Looking at the blame history, `cancelOnError` was set to true from the beginning, it's not clear why this was, but it seems like there is no other unanticipated side effect of this change.

This was manually tested on the web platform that websockets are now cleanly closed without exception, that no reconnect attempt is ever made during an explicit dispose of the `WebSocketLink`, and that reconnection attempts will continue to happen (if configured) until a connection is successfully made.

This might help with https://github.com/zino-hofmann/graphql-flutter/issues/873 but I don't know for sure if the issue described there is for sure related to any of the above.

vincenzopalazzo commented on packages/graphql/lib/src/links/websocket_link/websocket_client.dart:
```diff
@@ -173,6 +173,7 @@ class SocketClient {
       HashMap();
 
   bool _connectionWasLost = false;
+  bool _wasDisposed = false;
```
Why don't add a new enum state?

we could improve also the state pattern, to make more smart changes of state around, but I put here the question that I had during the reading of your PR.

klondikedragon replied:
You mean another enum state for `_connectionStateController`? one of the potential challenges with that is that the BehaviorSubject won't allow changing the value of it (which has to happen by adding the value to the stream) if the subject (stream controller) has already been closed. since the `_connectionStateController` is only explicitly closed in one place right now, it could ensure that it's set to this new enum value prior to closure, and that nothing else would ever change the state away from that before changing the state. That would result in a lot of other guard conditions where the state is modified throughout the rest of the file, so it seemed more complex. I'm happy to try that approach though if you prefer.

vincenzopalazzo replied:
No, I like your approach, I was thinking that maybe it is possible to make the state pattern more "pattern" and not with an enum, but this is out of context of this PR.

Thanks

vincenzopalazzo commented on packages/graphql/lib/src/links/websocket_link/websocket_client.dart:
```diff
@@ -234,7 +247,9 @@ class SocketClient {
         onMessage,
         onDone: onConnectionLost,
-        cancelOnError: true,
+        // onDone will not be triggered if the subscription is
+        // auto-cancelled on error; make sure to pass false
+        cancelOnError: false,
```
For now, I have no in mind what is the use case that this change can not be wrong, but call `onDone` after an error sound wired to me.

can you explain what is the reason here?

klondikedragon replied:
Right now the reconnection logic happens in the `onDone` callback (within `onConnectionLost`) and so if `onDone` is not called then nothing ever gets notified there is an end to the websocket connection, and so no reconnection attempt is made ever again. In this context, it seems `onDone` just refers to the end of a stream, and in dart a stream can receive multiple errors without the stream being done (in this particular stream it seems that it will normally only get one error). 

said another way, without this change, here is the current behavior:
* a websocket connection is closed by the server
* onStreamError callback is called (which looks to normally be set to _defaultOnStreamError), which just logs the exception (`WebSocketChannelException`, but doesn't do anything else), and then the stream subscription is automatically canceled, so future stream events (e.g., stream "done" state) are not sent to this stream subscription
* nothing else happens, so websocket remains disconnected forever, with no future connection attempt, even though autoReconnect is true.

More changes could be put into `_defaultOnStreamError`, but because that is overrideable, it seems better just to keep the reconnection logic in the `onConnectionLost` callback which is called at the end of the stream (which will be reached whenever the websocket connection closes for any reason).

vincenzopalazzo replied:
Yeah, so we can stay with this design for now, and see if in the next major release we could change some things to make this exception handling more explicative

vincenzopalazzo reviewed (commented):
Thank looks into that I was scheduler to look inside this issue this week, but it is better that you start to work on this.

However, I have no strong disagreement, but I have some questions that I would like to discuss.

P.S: In addition, I introduced in https://github.com/zino-hofmann/graphql-flutter/pull/1010 the web socket inside the test, could you write some tests?

Thanks for your time.

klondikedragon commented:
@vincenzopalazzo - thanks, added some comments to try and explain thinking, see what you think. I've also added 3 more websocket tests. The `disconnect via dispose` test fails without the changes and passes with the changes. The other 2 tests provide more coverage.

In writing the `resubscribe after server disconnect` test, I discovered why it was not obvious that autoReconnect was not working previously if the server forcefully disconnects the websocket... In the test, it is only possible to gracefully close the websocket (either having the websocket client initiate the close, or the websocket server initiates the close -- either way the websocket is shutdown gracefully and the socketChannel stream ends and the `onDone` event is sent, triggering the reconnect).

One behavior this PR fixes is the case of a forceful disconnect where the server just closes the socket without sending a graceful disconnect payload. In this case, instead of the socketChannel stream getting an `onDone` message, it gets an error injected into the stream (`WebSocketChannelException`). This can be common if the client's internet connection drops while the websocket is connected, or if the server / server process crashes, etc.

I couldn't find a way to test this particular case easily with a dart unit test, as the socketChannel does not expose its stream controller and so an error cannot easily be injected into it, and there doesn't seem to be a way to forcefully close the websocket non-gracefully (this is good that dart is enforcing this in its library interface for websockets, but makes it hard to test this error path). While not perfectly comprehensive, hopefully this PR improves websocket's resiliency incrementally without making it any worse and could be improved on later as needed.

vincenzopalazzo commented on packages/graphql/test/websocket_test.dart:
```diff
@@ -159,6 +165,47 @@ Future<void> main() async {
+          await socketClient.dispose();
+        });
+      })();
+      // The connectionState BehaviorController emits the current state
```
```suggestion
     
       // The connectionState BehaviorController emits the current state
```
I preferer a space here

vincenzopalazzo reviewed (changes requested):
I'm pretty happy with your work and thanks for doing this, I appreciate your edition to fix this issue and to improve the test case on the web socket, but also into describing the PR, this is how a PR needs to look like.

I have only a small request, and before merging this I need to catch why the `codcov` is failing, now I hope in this day I will push the change in this PR and merge it.

In addition, I would like to know also the opinion of @budde377 about this PR.

vincenzopalazzo reviewed (approved):
LGTM, thanks for your good job.
